src/components/registration-form.js

```diff
@@ -1,148 +1,285 @@
-import { SAVE_USER_RESULT } from './../domain/constants.js';
 import DisplayNameFormItem from './displayName-form-item.js';
+import { useDateFormat, useLocale } from './locale-context.js';
 import UserApiClient from './../api-clients/user-api-client.js';
-import PasswordConfirmationFormItem from './password-confirmation-form-item.js';
+import { Form, Button, Checkbox, Input, Steps, Result } from 'antd';
+import { LoadingOutlined, SmileOutlined, SolutionOutlined, UserOutlined } from '@ant-design/icons';
+import { HTTP_STATUS, PENDING_USER_REGISTRATION_EXPIRATION_IN_MINUTES, SAVE_USER_RESULT } from './../domain/constants.js';
 
 const logger = new Logger(import.meta.url);
 
 const FormItem = Form.Item;
 
-function RegistrationForm({
-  formRef,
-  onRegistrationFailed,
-  onRegistrationStarted,
-  onRegistrationSucceeded
-}) {
-  const [form] = Form.useForm();
+const STEP = {
+  enterData: 'enterData',
+  enterCode: 'enterCode',
+  completed: 'completed'
+};
+
+const STEP_STATE = {
+  wait: 'wait',
+  process: 'process',
+  finish: 'finish',
+  error: 'error'
+};
+
+const createDefaultSteps = () => [
+  {
+    key: STEP.enterData,
+    status: STEP_STATE.wait,
+    icon: <UserOutlined />,
+    translationKey: 'step_enterData'
+  },
+  {
+    key: STEP.enterCode,
+    status: STEP_STATE.wait,
+    icon: <SolutionOutlined />,
+    translationKey: 'step_enterCode'
+  },
+  {
+    key: STEP.completed,
+    status: STEP_STATE.wait,
+    icon: <SmileOutlined />,
+    translationKey: 'step_completed'
+  }
+];
+
+const setStepError = (steps, key) => steps.map(step => step.key === key ? { ...step, status: STEP_STATE.error } : step);
+const setStepSuccess = (steps, key) => steps.map(step => step.key === key ? { ...step, status: STEP_STATE.finish } : step);
+
+function RegistrationForm() {
+  const setUser = useSetUser();
   const settings = useSettings();
   const { uiLanguage } = useLocale();
-  const [user, setUser] = useState(null);
+  const [enterDataForm] = Form.useForm();
+  const [enterCodeForm] = Form.useForm();
+  const { formatDuration } = useDateFormat();
   const userApiClient = useService(UserApiClient);
   const { t } = useTranslation('registrationForm');
+  const [isLoading, setIsLoading] = useState(false);
+  const [createdUser, setCreatedUser] = useState(null);
+  const [steps, setSteps] = useState(createDefaultSteps());
   const [forbiddenEmails, setForbiddenEmails] = useState([]);
+  const [currentStepKey, setCurrentStepKey] = useState(STEP.enterData);
+  const [invalidVerificationCodes, setInvalidVerificationCodes] = useState([]);
 
-  if (formRef) {
-    formRef.current = form;
-  }
+  useEffect(() => {
+    if (currentStepKey === STEP.enterData) {
+      setCreatedUser(null);
+      setForbiddenEmails([]);
+      setInvalidVerificationCodes([]);
+    }
+    setSteps(previousSteps => {
+      const newSteps = currentStepKey === STEP.enterData ? createDefaultSteps() : previousSteps.map(step => ({ ...step }));
```

Comment:
For the sake of better clarity I would suggest to create a `setStepProcess` method just like `setStepError` and `setStepSuccess`. That pattern makes it so much easier to follow these state changes.

Reply:
Good idea, simplified the function accordingly.